### PR TITLE
Uf 4527 create scheduled job

### PR DESCRIPTION
**Pull request by MickeWallner**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -44,4 +44,4 @@ hs_err_pid*
 /target/
 
 # Local environment
-.env
```
Comment (josc04) [inline .gitignore:47]: Keep .env to ensure that no old .env files makes their way to Github?

File: pom.xml [hunk 2]
```diff
@@ -117,10 +122,20 @@
 					<sources>
 						<source>src/main/resources/schema/InExchangeInvoiceStatusSchema.xsd</source>
 					</sources>
-					<packageName>generated.se.inexchange</packageName>
+					<packageName>se.inexchange.generated</packageName>
```
Comment (josc04) [inline pom.xml:125]: I changed this as I though it was cleaner to have all generated classes under the same root package (generated). Don't you agree?

File: src/main/java/se/sundsvall/seabloader/api/InvoicesResource.java [hunk 3]
```diff
@@ -43,4 +42,14 @@ public ResponseEntity<Void> createInvoice(@NotNull @RequestBody final byte[] bod
 		invoiceService.create(body);
 		return noContent().build();
 	}
+
+	@PostMapping(path = "/exports", produces = APPLICATION_PROBLEM_JSON_VALUE)
```
Comment (josc04) [inline src/main/java/se/sundsvall/seabloader/api/InvoicesResource.java:46]: Move this to own resource class (/schedulers or likewise) to separate (and possibly also hide) resources that maybe should not be available to all clients?

File: src/main/java/se/sundsvall/seabloader/api/model/InvoiceType.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+	INVOICE("Faktura"),
+	CREDIT_INVOICE("Kreditfaktura"),
+	DIRECT_DEBIT("Autogiro"),
+	SELF_INVOICE("Självfaktura"),
+	REMINDER("Påminnelse"),
+	CONSOLIDATED_INVOICE("Samlingsfaktura");
+
+	private final String value;
+
+	InvoiceType (final String v) {
```
Comment (josc04) [inline src/main/java/se/sundsvall/seabloader/api/model/InvoiceType.java:13]: private?

File: src/main/java/se/sundsvall/seabloader/api/model/InvoiceType.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+		value = v;
+	}
+
+	public String value() {
+		return value;
+	}
+
+	public static InvoiceType fromValue(final String value) {
+		for (InvoiceType invoiceType : InvoiceType.values()) {
```
Comment (josc04) [inline src/main/java/se/sundsvall/seabloader/api/model/InvoiceType.java:22]: To use streams, just do this:
```
return java.util.Arrays.stream(values())
	.filter(enumObj -> enumObj.value.equals(value))
	.findFirst().orElseThrow(() -> new IllegalArgumentException(format("Illegal enum value: '%s'", stadsbackenValue)));

```

File: src/main/java/se/sundsvall/seabloader/integration/db/InvoiceRepository.java [hunk 6]
```diff
@@ -33,6 +32,15 @@ public interface InvoiceRepository extends JpaRepository<InvoiceEntity, Long>, J
 	 */
 	boolean existsByInvoiceId(String invoiceId);
 
+	/**
+	 * Find by status list.
+	 *
+	 * @param statusList a List of statuses to filter on.
+	 * @return A List of invoiceIds.
+	 */
+	@Query("SELECT invoiceId FROM InvoiceEntity i where i.status in ?1")
```
Comment (josc04) [inline src/main/java/se/sundsvall/seabloader/integration/db/InvoiceRepository.java:41]: This can be done with projections instead of using native queries. See PR https://github.com/Sundsvallskommun/api-service-seab-loader/pull/21 for how its done (with projections and a interface)

File: pom.xml [hunk 7]
```diff
@@ -174,34 +189,6 @@
 					</execution>
 				</executions>
 			</plugin>
-			<plugin>
-				<groupId>org.codehaus.mojo</groupId>
```
Comment (foobar-1337) [inline pom.xml:178]: Why is this removed?

File: src/main/java/se/sundsvall/seabloader/integration/db/InvoiceRepository.java (hunk 6 above)
Comment (foobar-1337) [inline src/main/java/se/sundsvall/seabloader/integration/db/InvoiceRepository.java:41]: Why the need for Query here? Use projections instead.

File: src/main/java/se/sundsvall/seabloader/integration/invoicecache/InvoiceCacheClient.java [hunk 8]
```diff
@@ -18,6 +19,6 @@ public interface InvoiceCacheClient {
 	 * 
 	 * @param invoicePdfRequest with attributes for export an invoice.
 	 */
-	@PostMapping(path = "importInvoice", produces = APPLICATION_PROBLEM_JSON_VALUE)
+	@PostMapping(path = "invoices", consumes = APPLICATION_JSON_VALUE, produces = APPLICATION_PROBLEM_JSON_VALUE)
 	ResponseEntity<Void> importInvoice(@RequestBody InvoicePdfRequest invoicePdfRequest);
```
Comment (foobar-1337) [inline src/main/java/se/sundsvall/seabloader/integration/invoicecache/InvoiceCacheClient.java:23]: I think that the method name in the client should be constructed from the view-point of this service.

(This service doesn't import anything. It sends/stores/trasmits invoices.)

File: src/main/java/se/sundsvall/seabloader/service/InvoiceService.java [hunk 9]
```diff
@@ -1,23 +1,37 @@
...
+import static se.sundsvall.seabloader.integration.db.model.enums.Status.UNPROCESSED;
+import static se.sundsvall.seabloader.service.mapper.InvoiceMapper.toInExchangeInvoice;
+import static se.sundsvall.seabloader.service.mapper.InvoiceMapper.toInvoiceEntity;
+import static se.sundsvall.seabloader.service.mapper.InvoiceMapper.toInvoicePdfRequest;
 
 @Service
 public class InvoiceService {
 
 	private static final Logger LOGGER = LoggerFactory.getLogger(InvoiceService.class);
+	private static final Status[] STATUSES_TO_SEND = {UNPROCESSED, FAILED};
```
Comment (josc04) [inline src/main/java/se/sundsvall/seabloader/service/InvoiceService.java:27]: I think there was discussion of having a separate status for incoming requests that failed to be processed and for those that failed when sending them to invoice cache, right? Now all failing entities will be retried and sent to invoice cache, but those that was unprocessable upon receival will never be successfully sent to cache.

TLDR; Add new status, EXPORT_FAILED, which will be set when export is not successful. Replace FAILED with the new status in the list above.

File: src/main/java/se/sundsvall/seabloader/service/InvoiceService.java [hunk 10]
```diff
@@ -29,4 +43,41 @@ public void create(final byte[] content) {
...
+		invoiceIds.forEach(invoiceId ->
+			invoiceRepository.findByInvoiceId(invoiceId)
+				.ifPresent(invoiceEntity -> {
+					try {
+						final var invoicePdfRequest = toInvoicePdfRequest(toInExchangeInvoice(invoiceEntity.getContent()));
+						invoiceCacheClient.importInvoice(invoicePdfRequest);
+						invoiceEntity.setStatus(PROCESSED);
+						invoiceEntity.setContent(null);
+						invoiceRepository.save(invoiceEntity);
+						LOGGER.info("Invoice {} sent to Invoice-cache", invoicePdfRequest);
```
Comment (josc04) [inline src/main/java/se/sundsvall/seabloader/service/InvoiceService.java:75]: This will create a lot of logs...

File: src/main/java/se/sundsvall/seabloader/service/InvoiceService.java [hunk 11]
```diff
@@ -29,4 +43,41 @@ public void create(final byte[] content) {
...
+
+	private void sendInvoiceToInvoiceCache(final List<String> invoiceIds) {
+		invoiceIds.forEach(invoiceId ->
+			invoiceRepository.findByInvoiceId(invoiceId)
+				.ifPresent(invoiceEntity -> {
+					try {
+						final var invoicePdfRequest = toInvoicePdfRequest(toInExchangeInvoice(invoiceEntity.getContent()));
+						invoiceCacheClient.importInvoice(invoicePdfRequest);
+						invoiceEntity.setStatus(PROCESSED);
+						invoiceEntity.setContent(null);
```
Comment (foobar-1337) [inline src/main/java/se/sundsvall/seabloader/service/InvoiceService.java:73]: Why?

File: src/main/java/se/sundsvall/seabloader/service/InvoiceService.java [hunk 12]
```diff
@@ -29,4 +43,41 @@ public void create(final byte[] content) {
...
+						invoiceRepository.save(invoiceEntity);
+						LOGGER.info("Invoice {} sent to Invoice-cache", invoicePdfRequest);
+					} catch (Exception e) {
+						LOGGER.error("Could not parse invoice with invoiceId: {} {}", invoiceId, e.getMessage());
```
Comment (foobar-1337) [inline src/main/java/se/sundsvall/seabloader/service/InvoiceService.java:77]: Are you sure that it's the parsing that will fail here?

File: src/main/java/se/sundsvall/seabloader/service/InvoiceService.java [hunk 13]
```diff
@@ -29,4 +43,41 @@ public void create(final byte[] content) {
...
+						invoiceEntity.setStatus(FAILED);
```
Comment (foobar-1337) [inline src/main/java/se/sundsvall/seabloader/service/InvoiceService.java:78]: We have a statusMessage field as well. I would suggest to populate this field with useful information.

File: src/main/java/se/sundsvall/seabloader/service/mapper/InvoiceMapper.java [hunk 14]
```diff
@@ -57,4 +74,30 @@ public static InExchangeInvoiceStatusType toInExchangeInvoice(final String xml)
 		final var unmarshaller = JAXBContext.newInstance(InExchangeInvoiceStatusType.class).createUnmarshaller();
 		return (InExchangeInvoiceStatusType) JAXBIntrospector.getValue(unmarshaller.unmarshal(xmlSource));
 	}
+
+	private static InvoicePdf toInvoicePdf(final InExchangeInvoiceStatusType inExchangeInvoiceStatusType) {
```
Comment (josc04) [inline src/main/java/se/sundsvall/seabloader/service/mapper/InvoiceMapper.java:78]: The InvoicePdfMerger does this already, no need to have all this logic here. Why not just autowire in the InvoicePdfMerger (it is written as a Spring component), have the initial null check on both **originalInvoice** and **attachments** and then call mergePdfs with inExchangeInvoiceStatusType as argument, and finally return a InvoicePdf with name from inExchangeInvoice.getOriginalinvoice().getName() and the response from the merger encoded as Base64 string?

File: src/main/resources/application.properties [hunk 15]
```diff
@@ -45,8 +45,6 @@ integration.invoicecache.readTimeout=20
 # Oauth2
 spring.security.oauth2.client.registration.invoicecache.authorization-grant-type=client_credentials
 spring.security.oauth2.client.registration.invoicecache.provider=invoicecache
-# Circuit breaker settings
```
Comment (josc04) [inline src/main/resources/application.properties:48]: IMHO I think it's easier to keep all settings for the integration together in one place

File: src/main/java/se/sundsvall/seabloader/service/mapper/InvoiceMapper.java [hunk 16]
```diff
@@ -57,4 +74,30 @@ public static InExchangeInvoiceStatusType toInExchangeInvoice(final String xml)
...
+			.map(InExchangeInvoiceStatusTypeAttachment.Attachment::getName)
+			.orElse(null);
+
+		if (Objects.isNull(name)) {
+			LOGGER.warn("No attachments found in invoice with invoiceId: {}", inExchangeInvoiceStatusType.getInvoice().getInvoiceId());
+			return null;
+		}
+
+		return new InvoicePdf()
+			.content(Base64.getEncoder().encodeToString(invoicePdfMerger.mergePdfs(inExchangeInvoiceStatusType).toString().getBytes(Charset.defaultCharset())))
```
Comment (foobar-1337) [inline src/main/java/se/sundsvall/seabloader/service/mapper/InvoiceMapper.java:100]: Is `toString()` really the correct way to transform an outputStream into a String?
Have you tested this?

File: src/main/java/se/sundsvall/seabloader/service/InvoiceService.java (hunk 11 above)
Comment (MickeWallner) [reply src/main/java/se/sundsvall/seabloader/service/InvoiceService.java:73]: We said last week that we could remove content after invoice is sent.

File: src/main/java/se/sundsvall/seabloader/service/InvoiceService.java (hunk 9 above)
Comment (MickeWallner) [reply src/main/java/se/sundsvall/seabloader/service/InvoiceService.java:27]: I've missed this discussion. I'll add the new status.

File: src/main/java/se/sundsvall/seabloader/service/InvoiceService.java (hunk 11 above)
Comment (foobar-1337) [reply src/main/java/se/sundsvall/seabloader/service/InvoiceService.java:73]: Why?

File: src/main/resources/application.properties [hunk 17]
```diff
@@ -23,9 +23,9 @@ openapi.version=@project.version@
 # Sort operations by method
 springdoc.swagger-ui.operationsSorter=method
 
-#===============================================================================
+#========================================
```
Comment (foobar-1337) [inline src/main/resources/application.properties:26]: The previous version (length) was better, since had the same lenght as all other dividers.

File: src/main/java/se/sundsvall/seabloader/service/InvoiceService.java (hunk 11 above)
Comment (foobar-1337) [reply src/main/java/se/sundsvall/seabloader/service/InvoiceService.java:73]: The space is already reserved. It will not help to set it to null, unless you also runs ` OPTIMIZE TABLE`.
I can't see that this is occuring. Hence: this is a useless action to perform.

Why not rely on the "database cleanup"-mechanizm instead?

File: src/main/java/se/sundsvall/seabloader/api/model/InvoiceType.java (hunk 4 above)
Comment (MickeWallner) [reply src/main/java/se/sundsvall/seabloader/api/model/InvoiceType.java:13]: Is redundant of enum constructors.

File: src/main/java/se/sundsvall/seabloader/service/InvoiceService.java (hunk 9 above)
Comment (foobar-1337) [inline src/main/java/se/sundsvall/seabloader/service/InvoiceService.java:27]: Rename to STATUSES_OF_INVOICES_TO_SEND (or something like that)?

(we are not sending statuses)

File: src/main/java/se/sundsvall/seabloader/service/InvoiceService.java [hunk 18]
```diff
@@ -29,4 +43,41 @@ public void create(final byte[] content) {
...
+		final var invoiceIdsToSend = invoiceRepository.findInvoiceIdsByStatusIn(STATUSES_TO_SEND);
+
+		if (invoiceIdsToSend.isEmpty()) {
+			LOGGER.info("No invoices found with status {}", STATUSES_TO_SEND);
+			return;
+		}
+
+		LOGGER.info("Found {} invoices with status {}", invoiceIdsToSend.size(), STATUSES_TO_SEND);
+
+		new ArrayList<>(ListUtils
```
Comment (foobar-1337) [inline src/main/java/se/sundsvall/seabloader/service/InvoiceService.java:60]: What is the point of sending this 10 by 10?
Why not just loop through all invoices (i.e invoiceIdsToSend) and send them 1 by 1?

File: src/main/java/se/sundsvall/seabloader/api/model/InvoiceType.java (hunk 4 above)
Comment (josc04) [reply src/main/java/se/sundsvall/seabloader/api/model/InvoiceType.java:13]: Yes of course, my bad

File: src/main/java/se/sundsvall/seabloader/service/InvoiceService.java (hunk 11 above)
Comment (MickeWallner) [reply src/main/java/se/sundsvall/seabloader/service/InvoiceService.java]: Fixed

File: src/main/java/se/sundsvall/seabloader/service/mapper/InvoiceMapper.java (hunk 16 above)
Comment (MickeWallner) [reply src/main/java/se/sundsvall/seabloader/service/mapper/InvoiceMapper.java]: Yes, and it seems to work.

File: src/main/java/se/sundsvall/seabloader/integration/db/model/enums/Status.java [hunk 19]
```diff
@@ -3,5 +3,6 @@
 public enum Status {
 	UNPROCESSED,
 	PROCESSED,
+	EXPORT_FAILED,
 	FAILED
```
Comment (josc04) [inline src/main/java/se/sundsvall/seabloader/integration/db/model/enums/Status.java:7]: Should we rename this to IMPORT_FAILED to better describe the status?

File: src/main/java/se/sundsvall/seabloader/service/mapper/InvoiceMapper.java [hunk 20]
```diff
@@ -57,4 +72,16 @@ public static InExchangeInvoiceStatusType toInExchangeInvoice(final String xml)
 		final var unmarshaller = JAXBContext.newInstance(InExchangeInvoiceStatusType.class).createUnmarshaller();
 		return (InExchangeInvoiceStatusType) JAXBIntrospector.getValue(unmarshaller.unmarshal(xmlSource));
 	}
+
+	private static InvoicePdf toInvoicePdf(final InExchangeInvoiceStatusType inExchangeInvoiceStatusType, final OutputStream outputStream) {
+
+		if (Objects.isNull(inExchangeInvoiceStatusType.getOriginalInvoice()) || Objects.isNull(outputStream)) {
+			LOGGER.warn("OriginalInvoice or attachments not found in invoice with invoiceId: {}", inExchangeInvoiceStatusType.getInvoice().getInvoiceId());
+			return null;
```
Comment (josc04) [inline src/main/java/se/sundsvall/seabloader/service/mapper/InvoiceMapper.java:80]: As it is now,all requests will be sent, even those that have no attachments. Is it verified that InvoiceCache should be called even if attachments are missing?

An alternative is to throw an exception (causing the entity/dbpost to fail and be marked as EXPORT_FAILED in the DB) or that the service layer have logic to not send requests containing no attachments to InvoiceCache.

File: src/main/java/se/sundsvall/seabloader/integration/db/model/enums/Status.java (hunk 19 above)
Comment (foobar-1337) [reply src/main/java/se/sundsvall/seabloader/integration/db/model/enums/Status.java:7]: Isn't this the failure status for the export (invoiceCache)?

File: src/main/java/se/sundsvall/seabloader/service/mapper/InvoiceMapper.java (hunk 20 above)
Comment (MickeWallner) [reply src/main/java/se/sundsvall/seabloader/service/mapper/InvoiceMapper.java:80]: I'll check this.

File: src/main/java/se/sundsvall/seabloader/integration/db/model/enums/Status.java (hunk 19 above)
Comment (MickeWallner) [reply src/main/java/se/sundsvall/seabloader/integration/db/model/enums/Status.java:7]: I think EXPORT_FAILED is better. It could be depend on parsing failed and then import is'nt in question.

File: src/main/java/se/sundsvall/seabloader/service/mapper/InvoiceMapper.java (hunk 20 above)
Comment (MickeWallner) [reply src/main/java/se/sundsvall/seabloader/service/mapper/InvoiceMapper.java:80]: This should result in an error according to Jakob. I'll fix.

File: src/main/java/se/sundsvall/seabloader/service/InvoiceService.java [hunk 21]
```diff
@@ -70,6 +71,10 @@ private void sendInvoiceToInvoiceCache(final List<InvoiceId> invoiceIds) {
 						invoiceCacheClient.sendInvoice(invoicePdfRequest);
 						invoiceEntity.setStatus(PROCESSED);
 						invoiceRepository.save(invoiceEntity);
+					} catch (ThrowableProblem tb) {
```
Comment (foobar-1337) [inline src/main/java/se/sundsvall/seabloader/service/InvoiceService.java:74]: What is the difference with the catch block below (except the log message)?

File: src/main/java/se/sundsvall/seabloader/service/mapper/InvoiceMapper.java [hunk 22]
```diff
@@ -76,8 +79,8 @@ public static InExchangeInvoiceStatusType toInExchangeInvoice(final String xml)
 	private static InvoicePdf toInvoicePdf(final InExchangeInvoiceStatusType inExchangeInvoiceStatusType, final OutputStream outputStream) {
 
 		if (Objects.isNull(inExchangeInvoiceStatusType.getOriginalInvoice()) || Objects.isNull(outputStream)) {
-			LOGGER.warn("OriginalInvoice or attachments not found in invoice with invoiceId: {}", inExchangeInvoiceStatusType.getInvoice().getInvoiceId());
-			return null;
+			LOGGER.error("OriginalInvoice or attachments not found in invoice with invoiceId: {}", inExchangeInvoiceStatusType.getInvoice().getInvoiceId());
+			throw Problem.valueOf(BAD_REQUEST, format("OriginalInvoice or attachments not found in invoice with invoiceId: %s", inExchangeInvoiceStatusType.getInvoice().getInvoiceId()));
```
Comment (foobar-1337) [inline src/main/java/se/sundsvall/seabloader/service/mapper/InvoiceMapper.java:83]: Why are you using a Problem with a BAD_REQUEST-status (client error) here? No status is sent back to any client right?

Why not just an IllegalArgumentException?

File: src/main/java/se/sundsvall/seabloader/service/InvoiceService.java (hunk 21 above)
Comment (MickeWallner) [reply src/main/java/se/sundsvall/seabloader/service/InvoiceService.java:74]: No difference more than avoiding an extra log-message. I'll change it to only catch Exception.
